llvm-external-projects/iree-dialects/lib/Transforms/ListenerGreedyPatternRewriteDriver.cpp

```diff
@@ -438,31 +401,249 @@ LogicalResult GreedyPatternRewriteDriver::notifyMatchFailure(
+/// This driver simplfies a list of ops.
+class MultiOpPatternRewriteDriver : public GreedyPatternRewriteDriver {
+public:
+  explicit MultiOpPatternRewriteDriver(
+      RewriteListener *listener, MLIRContext *ctx,
+      const FrozenRewritePatternSet &patterns,
+      const GreedyRewriteConfig &config, ArrayRef<Operation *> ops,
+      llvm::SmallDenseSet<Operation *, 4> *survivingOps = nullptr);
+
+  /// Simplify `ops`. Return `success` if the transformation converged.
+  LogicalResult simplify(ArrayRef<Operation *> ops, bool *changed = nullptr) &&;
+
+private:
+  void notifyOperationRemoved(Operation *op) override {
+    GreedyPatternRewriteDriver::notifyOperationRemoved(op);
+    if (survivingOps)
+      survivingOps->erase(op);
+  }
+
+  /// An optional set of ops that survived the rewrite. This set is populated
+  /// at the beginning of `simplifyLocally` with the inititally provided list
+  /// of ops.
+  llvm::SmallDenseSet<Operation *, 4> *const survivingOps = nullptr;
+};
+} // namespace
+
+MultiOpPatternRewriteDriver::MultiOpPatternRewriteDriver(
+    RewriteListener *listener, MLIRContext *ctx,
+    const FrozenRewritePatternSet &patterns, const GreedyRewriteConfig &config,
+    ArrayRef<Operation *> ops,
+    llvm::SmallDenseSet<Operation *, 4> *survivingOps)
+    : GreedyPatternRewriteDriver(listener, ctx, patterns, config),
+      survivingOps(survivingOps) {
+  if (config.strictMode != GreedyRewriteStrictness::AnyOp)
+    strictModeFilteredOps.insert(ops.begin(), ops.end());
+
+  if (survivingOps) {
+    survivingOps->clear();
+    survivingOps->insert(ops.begin(), ops.end());
+  }
+}
+
+LogicalResult MultiOpPatternRewriteDriver::simplify(ArrayRef<Operation *> ops,
+                                                    bool *changed) && {
+  // Populate the initial worklist.
+  for (Operation *op : ops)
+    addSingleOpToWorklist(op);
+
+  // Process ops on the worklist.
+  bool result = processWorklist();
+  if (changed)
+    *changed = result;
+
+  return success(worklist.empty());
+}
+
+/// Find the region that is the closest common ancestor of all given ops.
+static Region *findCommonAncestor(ArrayRef<Operation *> ops) {
+  assert(!ops.empty() && "expected at least one op");
+  // Fast path in case there is only one op.
+  if (ops.size() == 1)
+    return ops.front()->getParentRegion();
+
+  Region *region = ops.front()->getParentRegion();
+  ops = ops.drop_front();
+  int sz = ops.size();
+  llvm::BitVector remainingOps(sz, true);
+  do {
+    int pos = -1;
+    // Iterate over all remaining ops.
+    while ((pos = remainingOps.find_first_in(pos + 1, sz)) != -1) {
+      // Is this op contained in `region`?
+      if (region->findAncestorOpInRegion(*ops[pos]))
+        remainingOps.reset(pos);
+    }
+    if (remainingOps.none())
+      break;
+  } while ((region = region->getParentRegion()));
+  assert(region && "could not find common parent region");
+  return region;
+}
+
+LogicalResult mlir::applyOpPatternsTrackAndFold(
```

Comment:
This looks custom but not marked as such

Reply:
Added a comment to the beginning of the file that the RewriteListener is passed through all entry points until the GreedyPatternRewriteDriver class.